src/binary_to_text/binary_encoder.ml

```diff
@@ -0,0 +1,713 @@
+  | _ -> assert false
+
+let valtype_encode vt buf =
+  let c =
+    match vt with
+    | Num_type I32 -> '\x7F'
+    | Num_type I64 -> '\x7E'
+    | Num_type F32 -> '\x7D'
+    | Num_type F64 -> '\x7C'
+    | Ref_type (Null, Func_ht) -> '\x70'
+    | Ref_type (Null, Extern_ht) -> '\x6F'
+    | _ -> assert false (* vecttype v128 '\x7B' *)
+  in
+  add_char c buf
+
+let vector_encode datas func_encode buf =
+  let vector_buf = Buffer.create 16 in
+  let len =
+    List.fold_left
+      (fun acc data ->
+        match func_encode data vector_buf with
+        | None -> acc
+        | Some () -> acc + 1 )
+      0 datas
+  in
+  write_U32_of_int len buf;
+  Buffer.add_buffer buf vector_buf
+
+let vector_encode_from_namedvalues ~rev_values datas func_encode buf =
+  let { Named.values; _ } = datas in
+  match values with
+  | [] -> ()
+  | values ->
+    let values = if rev_values then List.rev values else values in
+    vector_encode values func_encode buf
+
+let resulttype_encode (rt : _ result_type) buf =
+  vector_encode rt (fun vt buf -> Some (valtype_encode vt buf)) buf
+
+let paramtype_encode (pt : _ param_type) buf =
+  let vt = List.map snd pt in
+  vector_encode vt (fun vt buf -> Some (valtype_encode vt buf)) buf
+
+let mut_encode mut buf =
+  let c = match mut with Const -> '\x00' | Var -> '\x01' in
+  add_char c buf
+
+let block_type_encode (typ : binary block_type) buf =
+  match typ with
+  | Bt_raw (None, ([], [])) -> add_char '\x40' buf
+  | Bt_raw (None, ([], [ vt ])) -> valtype_encode vt buf
+  | _ -> assert false (* TODO: types link *)
+
+let global_type_encode (typ : _ global_type) buf =
+  let mut, vt = typ in
+  valtype_encode vt buf;
+  mut_encode mut buf
+
+let limits_encode limits buf =
+  match limits with
+  | { min; max = None } ->
+    add_char '\x00' buf;
+    write_U32_of_int min buf
+  | { min; max = Some max } ->
+    add_char '\x01' buf;
+    write_U32_of_int min buf;
+    write_U32_of_int max buf
+
+let memarg_encode ({ offset; align } : memarg) buf =
+  write_U32 offset buf;
+  write_U32 align buf
+
+let memory_encode (mem : mem) buf =
+  let _so, limits = mem in
+  limits_encode limits buf
+
+let memory_import_encode (mem : limits Imported.t) buf =
+  let { Imported.modul; name; assigned_name = _; desc = limits } = mem in
+  add_bytes modul buf;
+  add_bytes name buf;
+  add_char '\x02' buf;
+  limits_encode limits buf
+
+let table_encode (table : _ table) buf =
+  let _so, (limits, (_nullable, heaptype)) = table in
+  reftype_encode heaptype buf;
+  limits_encode limits buf
+
+let table_import_encode (table : _ table_type Imported.t) buf =
+  let { Imported.modul; name; assigned_name = _; desc } = table in
+  let limits, (_nullable, heaptype) = desc in
+  add_bytes modul buf;
+  add_bytes name buf;
+  add_char '\x01' buf;
+  reftype_encode heaptype buf;
+  limits_encode limits buf
+
+let fc_encode i buf =
+  add_char '\xFC' buf;
+  write_U32_of_int i buf
+
+let rec instr_encode instr buf =
```

Comment:
You can define:

```ocaml
let add_char c = add_char buf c in (* after changing the order in your function *)
match instr with ...
```

And it will make your code much less cluttered here.